### DISCUSSION TITLE
HTTPS support in dev server

### DISCUSSION THREAD

**Discussion by benjaminwy** [withastro/roadmap#80] [Proposal]
We've decided to use Astro at my company for our homepage area/landing pages, and one snag I'm running into is (seemingly, correct me if I'm wrong) lack of HTTPS support in dev. This causes problems with our user authentication system, which we can work around, but it would be ideal if we could have our site running on something like, for example, https://site.test:3030.

With the following `astro.config.mjs`, the site will happily load on localhost:3000, but the `vite.server` config [appears to be ignored](https://github.com/withastro/astro/blob/69af658b00be0a3b1bb0eb11c2e480973a5a6301/packages/astro/src/core/dev/index.ts#L235):

```js
import fs from 'fs';

const cert = fs.readFileSync('./certs/cert.pem');
const key = fs.readFileSync('./certs/key.pem');

export default {
  // devOptions: {
  //   hostname: 'localhost',
  //   port: 3000,
  // },
  renderers: [
    "@astrojs/renderer-vue"
  ],
  vite: {
    server: {
      host: 'site.test',
      port: 3030,
      https: { key, cert }
    }
  }
};
```

Is this something you folks would be interesting in supporting? And if so, would the desired approach be to add an `https` parameter to `devOptions`?

**Comment by jonathantneal** [comment]
I would like to see this supported, as well. At a glance, I think this is a proposal worth doing.

Under the hood, as you’ve mentioned, it seems like it _should_ be supported thru [Vite](https://vitejs.dev/config/#server-https), and this proposal _might_ be a kind of bug report.

**Comment by jonathantneal** [comment]
### Early Feedback from the Community RFC Call (2022-01-25)

- Lots of positive feedback to support the idea of this feature.
- Implementation of `vite.server` is currently problematic, because Vite server is not used.
- 1 person recommended Astro use the same server for development-mode and preview-mode.
- 1 person recommended Astro switch to using the Vite server for both modes.
- Recommendation to @withastro/maintainers-core: warn about unused `vite` configs.

**Comment by jonathantneal** [comment]
Another request for this came up today. https://canary.discord.com/channels/830184174198718474/936231242649698364

**Reply by FredKSchott** [reply]
This is really interesting! It looks like SvelteKit uses the `configureServer` method that I'd mentioned in the call. https://github.com/sveltejs/kit/blob/5af1b92ff61add82ebc369cf83f53ca7594fafc9/packages/kit/src/core/dev/plugin.js

That makes a lot of sense: you get all of the Vite dev server goodies for free, and also don't need to manage those dependencies (like "connect") yourself!

I'd love for someone to prototype something like this  to see if it could work for Astro!

**Reply by matthewp** [reply]
@FredKSchott I believe the initial reason why we created our own server was so that we would be able to handle requests before they got to Vite. I think this is due to the fact that Vite will fallback to SPA routing on routes it doesn't understand.

SvelteKit seems to get around this limitation by removing the HTML plugin: https://github.com/sveltejs/kit/blob/5af1b92ff61add82ebc369cf83f53ca7594fafc9/packages/kit/src/core/dev/plugin.js#L325-L335

**Reply by jonathantneal** [reply]
If removing the HTML plugin would unblock using the vite server, it would also benefit plugin authors.

The HTML plugin causes problems for other plugins. It triggers duplicate runs thru the Vite hook system, as any import or asset has to run twice thru the `resolveId` and `load` hooks, along with the others. This forces very peculiar caching and prefix-adding strategies on plugins. The HTML plugin is particularly meddling because the `importerId` is lost to hooks during that run; e.g. strings won’t match up for caching, and now you don’t know what imported the asset (_making resolving `"./file.ext"` a pain_).

**Reply by FredKSchott** [reply]
Yup, it looks like they do some heavy-duty cleanup of the server in that plugin. If that would also solve bugs for us, that sounds like a win-win!

**Comment by FredKSchott** [comment]
Update: i believe that this is now possible in this PR: https://github.com/withastro/astro/pull/2494

**Reply by jonathantneal** [reply]
This PR has been merged. Thanks, @FredKSchott! Enjoy, @benjaminwy!

**Comment by jonathantneal** [comment]
:wave: A followup. I just tested this and `https: true` does not appear to work. The `https://localhost:3000` page would not display. The `http://localhost:3000` page would refresh constantly and infinitely. The shell would display the `http` URL regardless.

```js
// @ts-check
export default /** @type {import('astro').AstroUserConfig} */ ({
	renderers: [],
	vite: {
		server: {
			https: true
		}
	}
});
```

**Reply by FredKSchott** [reply]
I believe that you need more than just `https: true`. In my experience this also requires a key and a cert file. @benjaminwy I'd be curious if you followed these instructions if you'd have more luck: https://stackoverflow.com/a/69743888/4682119

I'd also try this on SvelteKit, since this appears to follow the same approach now that they do.

Lastly, I wonder if there's a Vite error being swallowed here, since I'm surprised that it just silently fails here.

**Reply by benjaminwy** [reply]
I'm thrilled that you guys have moved on this so quickly! Thanks a bunch, @FredKSchott and @jonathantneal!

Fred's right that we'll need to pass in a key and cert. But even before getting to that, I'm having some difficulty making the server run on the host and port I define in `vite.server`. There's a high chance this is user error, so I'll start from the beginning and detail each step I took.

First, I tried creating a new project that uses the latest pre-release (not sure if this is the way to go about it):

1. `yarn create astro`
2. `yarn add astro@next`  
(`yarn astro --version` yields `0.23.0-next.0`)

Then I played around with a few `astro.config.mjs` files:

#### 1. Runs on `http://127.0.0.1:3000/`

```javascript
export default {
  renderers: [],
  vite: {
    server: {
      host: 'site.test',
      port: 8888
    }
  }
};
```

#### 2. Runs on `http://127.0.0.1:7777`

```javascript
export default {
  devOptions: {
    hostname: 'site.test',
    port: 7777,
  },
  renderers: [],
  vite: {
    server: {
      host: 'site.test',
      port: 8888
    }
  }
};
```

#### 3. Errors out -- `address not available 127.0.0.2:3000` 

I expected `127.0.0.2` to not be available, but after the previous two configs, I didn't expect the `vite.server.host` param to be picked up at all.

```javascript
export default {
  renderers: [],
  vite: {
    server: {
      host: '127.0.0.2',
      port: 8888
    }
  }
};
```

I'll play around with this some more tomorrow.

**Reply by jonathantneal** [reply]
> I believe that you need more than just `https: true`

> Fred's right that we'll need to pass in a key and cert

Y’all, I would not have shared this without testing. The boolean for `https` is supposed to work.

From https://vitejs.dev/config/#server-https
> ### server.https
> - **Type:** `boolean | https.ServerOptions`

When I say something doesn’t work in Astro, I mean it. I can be quick to say something doesn’t work when there are caveats, but those caveats are ignored by most users and/or poorly documented. We don’t have caveats about restricting `https`, it totally works in Vite, and it’s definitely documented. 😅

See for yourself. Following https://vitejs.dev/guide/#scaffolding-your-first-vite-project
```
npm init vite@latest
```

After running the setup, you can test this thru the CLI without any configuration.

See for yourself. Following https://vitejs.dev/guide/#command-line-interface
> You can specify additional CLI options like `--port` or `--https`. For a full list of CLI options, run `npx vite --help` in your project.

```
npm run dev -- --https
```

It works. And the URL in the shell is displayed correctly. The API mirrors this.

See for yourself. Following https://vitejs.dev/guide/api-javascript.html#createserver
```js
import { createServer } from 'vite'

const server = await createServer({
	// any valid user config options, plus `mode` and `configFile`
	configFile: false,
	root: process.cwd(),
	server: {
		https: true,
		port: 1337
	}
})

await server.listen()

server.printUrls()
```

```
  > Local: https://localhost:1337/
  > Network: use `--host` to expose
```

**Reply by FredKSchott** [reply]
Ah you're right, it does work. I was misremembering this screen as not passable, which is what I see when I run the `npm run dev -- --https` command above:
<img width="1052" alt="Screen Shot 2022-02-01 at 10 27 42 PM" src="https://user-images.githubusercontent.com/622227/152104899-5d7e43bf-2281-4fdb-a790-d038c597dd05.png">

Jon I can't reproduce what you're seeing though. If I add this to my config file and nothing else, I see https://localhost:3000 in the shell and my browser is able to hit that URL (although with the same blocking banner as screenshotted above):

```diff
  },
+  vite: {
+    server: {
+      https: true,
+    },
+  },
};
```

@benjaminwy as for your host + port issue, I'm not sure! I can confirm that `devOptions.port` and `--port` are both working, but I don't have a test setup to test a custom host name. I'd love your help if you're able to test more and find the issue! devOptions should be taking priority over the vite options.

**Reply by jonathantneal** [reply]
We seem to be getting different errors. In Vite, my experience matches your screenshot. In Astro, tho our configurations are the same, here is the error I see:
<img width="560" alt="ERR_SSL_PROTOCOL_ERROR" src="https://user-images.githubusercontent.com/188426/152152501-de90ded4-cc36-4b2c-a9e1-bca8e84e74a0.png">
```
ERR_SSL_PROTOCOL_ERROR
```

And this one cannot be bypassed. Similarly, the shell does not show https:
<img width="680" alt="image" src="https://user-images.githubusercontent.com/188426/152152721-ec4af207-5664-4e1c-8dac-70c9e4b3015a.png">

This was originally tested in Astro v0.22.18 and now again in v0.22.20.

**Reply by benjaminwy** [reply]
Sorry Jon! I stand corrected. FWIW with `https: true` in Astro v0.23.0-next.0, I'm seeing the following:

<img width="1051" alt="image" src="https://user-images.githubusercontent.com/2089521/152194889-62b71214-fe6c-4b27-a7c0-1f07d43c2769.png">

<img width="498" alt="image" src="https://user-images.githubusercontent.com/2089521/152194948-25872f86-a4a4-4baf-81ca-5e39637d9398.png">

Full config:

```javascript
export default {
  devOptions: {
  },
  vite: {
    server: {
      host: 'localhost',
      port: 4000,
      https: true
    }
  },
  renderers: [
    "@astrojs/renderer-vue"
  ],
};
```

Still no luck setting the host name, which I need to figure out before the key & cert would be usable. Will report back when I have something. 

(can confirm `devOptions.port` and `--port` are both working for me too)

**Reply by FredKSchott** [reply]
@jonathantneal that's very strange. can you confirm that you're working off of the latest `main` branch, and that you've recently run `yarn build`? That's the exact behavior that I'd expect before the Vite move had been merged in.

**Reply by FredKSchott** [reply]
Ah wait, I think I've figured out the `host` issue (where port works but host does not)!

**Comment by natemoo-re** [comment]
This has landed in https://github.com/withastro/astro/pull/2494, so I'm closing this discussion! Thanks everyone!
